Python/suggestions.c

```diff
@@ -312,6 +312,38 @@ offer_suggestions_for_name_error(PyNameErrorObject *exc)
     return result;
 }
 
+static PyObject *
+offer_suggestions_for_import_error(PyImportErrorObject *exc)
+{
+    PyObject *mod_name = exc->name; // borrowed reference
+    PyObject *name = exc->name_from; // borrowed reference
+    if (name == NULL || mod_name == NULL || name == Py_None ||
+        !PyUnicode_CheckExact(name) || !PyUnicode_CheckExact(mod_name)) {
+        return NULL;
+    }
+
+    PyObject* mod = PyImport_Import(mod_name);
```

Comment:
I was wondering whether we could avoid "realy-importing" a module if it isn't already (e.g. whether it is currently cached or not)? (IIRC `from x import y` should put the `x` to the `sys.modules`, but might be missing something too).

Reply:
`PyImport_Import` goes through the cache if for whatever reason the module is already imported, no? (And indeed, the module is imported as part of raising the exception).

Reply:
> PyImport_Import goes through the cache if for whatever reason the module is already imported, no? (And indeed, the module is imported as part of raising the exception).

Exactly, but consider the scenario where it isn't already imported and the module itself has some side effects (like it prints something at the top level). Importing it would then might produce unexpected side effects that only appear when the traceback is printed, so I thought whether we could limit this hint when the module is already cached.

```
if module not in sys.modules: return None # give up if it isn't already in the cache
module = PyImport_Import(mod_name)
...
```

Reply:
> but consider the scenario where it isn't already imported and the module itself has some side effects

Hummmm. But isn't this a pre-condition for this kind of import error? If `from foo import x` fails to import `foo` then there is no lookup failure and we will never arrive here. And to be able to look `x` in `foo` then `foo` needs to be in `sys.modules`. Maybe I am missing something ... :(

Reply:
I think it's at least theoretically possible for the module to be removed by now, because some Python code may run between the time the error is thrown and the time this code runs. For example, in pythonrun.c we call print_exception_message() before print_exception_suggestions(), and that does a bunch of allocations and DECREFs that could trigger GC.

You could call https://docs.python.org/3/c-api/import.html#c.PyImport_GetModule instead to return the module only if it's already imported.